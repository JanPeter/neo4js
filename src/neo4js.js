import { v1 as neo4j } from 'neo4j-driver';
import Utils from './Utils';
import uuid from 'node-uuid';
import Debug from 'debug';
const debug = Debug('neo4js');

import * as neo4jsErrors from './errors';
import Model from './model';
import ModelManager from './model-manager';
import ModelObject from './model-object';
import Query from './query';
import { getSchemaInfo } from './schema-info';

class Neo4js {
  /**
   * Constructor to create a new instance of the Neo4js class
   * @param {String}  username The username which is used to authenticate against the database
   * @param {String}  password The password which is used to authenticate against the database
   * @param {Object}  [options={}] An object with options
   * @param {String}  [options.url=localhost] Bolt url
   * @param {Integer} [options.port=7687] Bolt port
   * @param {Object}  [options.rest={}]
   * @param {String}  [options.rest.url=localhost]
   * @param {Integer} [options.rest.port=7474]
   * @param {Integer} [options.connectionPoolSize=50]
   */
  constructor(username, password, options) {
    this.auth = { username, password };

    const defaultOptions = {
      url: `bolt://localhost`,
      port: 7687,
      rest: {
        url: `http://localhost`,
        port: 7474,
      },
      driver: {
        connectionPoolSize: 50,
      },
    };

    options = Utils._.extend({}, defaultOptions, options || {});
    options.rest = Utils._.extend({}, defaultOptions.rest, options.rest);
    options.driver = Utils._.extend({}, defaultOptions.driver, options.driver);

    // TODO: Maybe add here some more checks...
    const url = `${options.url}:${options.port}`;
    this.driver = neo4j.driver(url, neo4j.auth.basic(username, password), options.driver);

    this.options = options;

    this.models = {};
    this.modelManager = new ModelManager(this);
  }

  /**
   * @param {String | String[]} labels
   * @param {Object} schema
   * @param {Object} schema.{propertyName}
   * @param {Boolean} schema.{propertyName}.index
   * @param {Boolean} schema.{propertyName}.unique
   * @param {Boolean} schema.{propertyName}.exists
   * @param {Function} schema.{propertyName}.defaultValue
   */
  define(labels, schema, hooks) {
    const model = new Model(labels, schema, hooks, this);
    this.modelManager.add(model);
    return model;
  }

  /**
   * @param {String | String[]} labels
   */
  getModel(labels) {
    const model = this.modelManager.getModel(labels);
    return model;
  }

  /**
   * @returns {Promise}
   */
  getSchemaInfo() {
    return getSchemaInfo(this.options.rest.url, this.options.rest.port, this.auth);
  }

  drop() {
    return this.getSchemaInfo()
      .then(result => {
        const cmds = [];
        for (const label in result) {
          for (const p in result[label]) {
            if (result[label][p].index) {
              cmds.push(new Query(this).dropIndex(label, p));
            }
            if (result[label][p].unique) {
              cmds.push(new Query(this).dropConstraint(label, p, 'unique'));
            }
          }
        }
        cmds.push(new Query(this).match('n').detach('n'));
        return Query.runSequence(cmds);
      });
  }

  /**
   * @param {Object} options
   * @param {Boolean} options.force
   */
  sync(options) {
    debug(`sync start`);
    options = options || {};

    return Promise.resolve()
      .then(() => {
        if (options.force) {
          debug(`cleanup database start`);
          return this.drop();
        }
      })
      .then(() => {
        if (options.force) {
          debug(`cleanup database end`);
        }
        const promises = this.modelManager.all.map(m => m.sync());
        return Promise.all(promises);
      })
      .then(syncResults => {
        const errors = [];
        for (const results of syncResults) {
          for (const result of results) {
            if (result.err) {
              errors.push(result.err);
            }
          }
        }
        return errors;
      })
      .then(result => {
        debug(`sync end`);
        return result;
      })
      .catch(err => {
        debug(`sync end with errors`, err);
        return err;
      });
  }

  /**
   * This method creates a new instance of a Query
   * @return {Query}
   */
  Query() {
    return new Query(this);
  }

  /**
   * This method runs a command with params
   * @param {String} cmd - Cypher statement to execute
   * @param {Object} params - Map with parameters to use in statement
   * @return {Promise} Resolves either with the result or the error message
   */
  run(cmd, params) {
    const session = this.driver.session();
    return session.run(cmd, params)
      .then(result => {
        session.close();
        return result;
      })
      .catch(err => {
        session.close();
        return err;
      });
  }

  /**
   * @returns {UUIDv4}
   */
  static uuid4() {
    return uuid.v4();
  }

  /**
   * @returns {UUIDv1}
   */
  static uuid1() {
    return uuid.v1();
  }
}

Neo4js.prototype.Error = Neo4js.Error = neo4jsErrors.BaseError;
Neo4js.prototype.Model = Neo4js.Model = Model;
Neo4js.prototype.ModelObject = Neo4js.ModelObject = ModelObject;

export default Neo4js;
