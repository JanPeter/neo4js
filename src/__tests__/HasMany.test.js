// @flow

import neo4js, {
  Model,
  ModelInstance,
  src,
  model,
  dest,
  relation,
} from "../index";
import idx from "idx";

type PersonProps = {
  name?: StringProperty,
};

type TaskProps = {
  title?: StringProperty,
  done?: boolean,
};

class PersonModel extends Model<PersonProps, PersonInstance> {}
const Person: PersonModel = new PersonModel("Person");

class TaskModel extends Model<TaskProps, TaskInstance> {}
const Task: TaskModel = new TaskModel("Task");

const TaskCreatorRelation = relation("created").src
  .hasMany(Task)
  .dest.hasOne(Person);

const TaskAssigneeRelation = relation("assigned").src
  .hasMany(Task)
  .dest.hasOne(Person);

@model(Person)
class PersonInstance extends ModelInstance<PersonProps> {
  @src(TaskCreatorRelation) tasks: HasManyActions<TaskProps, TaskInstance>;

  @src(TaskAssigneeRelation)
  assignedTasks: HasManyActions<TaskProps, TaskInstance>;
}

@model(Task)
class TaskInstance extends ModelInstance<TaskProps> {
  @dest(TaskCreatorRelation)
  creator: HasOneActions<PersonProps, PersonInstance>;
}

describe("HasMany", () => {
  beforeAll(() => {
    neo4js.init({
      boltUri: "localhost",
      boltPort: 10001,
    });
  });

  beforeEach(async () => {
    await Person.create({ name: "Olaf" });
    await Person.create({ name: "Ignatz" });
    await Task.create({ title: "Learn magic", done: true });
    await Task.create({ title: "Write more test cases", done: false });
    await Task.create({ title: "Write more test cases", done: false });
  });

  afterEach(async () => {
    await neo4js.run("MATCH (n) DETACH DELETE n");
  });

  afterAll(() => {
    neo4js.close();
  });

  describe("create", () => {
    it("should create instances of tasks and relate them to a person instance", async () => {
      const paul: PersonInstance = await Person.create({ name: "Paul" });

      const propsArray: TaskProps[] = [
        {
          title: "Buy milk",
        },
        {
          title: "Buy beer",
          done: false,
        },
      ];

      const tasks: TaskInstance[] = await paul.tasks.create(propsArray);

      expect(
        tasks.map(t => {
          delete t.props.guid;
          return t;
        })
      ).toMatchSnapshot();
    });

    it("should create instances of tasks and relate them to a person instance with relation properties", async () => {
      const paul: PersonInstance = await Person.create({ name: "Paul" });

      const propsArray: TaskProps[] = [
        {
          title: "Buy milk",
        },
        {
          title: "Buy beer",
          done: false,
        },
      ];

      const tasks: TaskInstance[] = await paul.tasks.create(propsArray, {
        relationCreated: "today",
      });

      const db = await neo4js.run("MATCH (a)-[b]->(c) RETURN a, b, c");
      const data = db.map(item => {
        delete item.a.guid;
        delete item.c.guid;
        return { from: item.a, to: item.c, props: item.b };
      });

      expect(data).toMatchSnapshot();
    });

    it("should create instances of tasks and relate them to a person instance with multiple relation properties", async () => {
      const paul: PersonInstance = await Person.create({ name: "Paul" });

      const propsArray: TaskProps[] = [
        {
          title: "Buy milk",
        },
        {
          title: "Buy beer",
          done: false,
        },
      ];

      const tasks: TaskInstance[] = await paul.tasks.create(propsArray, {
        relationCreated: "today",
        type: "todo",
      });

      const db = await neo4js.run("MATCH (a)-[b]->(c) RETURN a, b, c");
      const data = db.map(item => {
        delete item.a.guid;
        delete item.c.guid;
        return { from: item.a, to: item.c, props: item.b };
      });

      expect(data).toMatchSnapshot();
    });
  });

  describe("get", () => {
    let paul: PersonInstance;

    const tasksProps: TaskProps[] = [
      {
        title: "Buy milk",
      },
      {
        title: "Buy beer",
        done: true,
      },
    ];

    beforeEach(async () => {
      paul = await Person.create({ name: "Paul" });
      const hubert: PersonInstance = await Person.create({ name: "Hubert" });
      hubert.tasks.create([{ title: "Learn to drive a car" }], {
        since: "today",
        type: "todo",
      });
    });

    it("should find all tasks to instance", async () => {
      await paul.tasks.create(tasksProps);
      const tasks: TaskInstance[] = await paul.tasks.get();

      expect(
        tasks.map(t => {
          delete t.props.guid;
          return t;
        })
      ).toMatchSnapshot();
    });

    it("should find all tasks to instance where task property done equals true", async () => {
      await paul.tasks.create(tasksProps);
      const tasks: TaskInstance[] = await paul.tasks.get({ done: true });

      expect(
        tasks.map(t => {
          delete t.props.guid;
          return t;
        })
      ).toMatchSnapshot();
    });

    it("should find all tasks to instance", async () => {
      await paul.tasks.create(tasksProps);
      const tasks: TaskInstance[] = await paul.tasks.get();

      expect(
        tasks.map(t => {
          delete t.props.guid;
          return t;
        })
      ).toMatchSnapshot();
    });

    it("should find nothing", async () => {
      await paul.tasks.create(tasksProps);
      const tasks: TaskInstance[] = await paul.assignedTasks.get();
      expect(tasks).toMatchSnapshot();
    });

    it("should find all tasks starting with character 'B' and are done", async () => {
      await paul.tasks.create(tasksProps);
      const tasks: TaskInstance[] = await paul.tasks.get({
        title: { $sw: "B" },
        done: true,
      });
      expect(
        tasks.map(t => {
          delete t.props.guid;
          return t;
        })
      ).toMatchSnapshot();
    });

    it("should find all tasks starting with character 'B' and are done with relationProp type todo", async () => {
      await paul.tasks.create(tasksProps);
      await paul.tasks.create([{ title: "Buy milk", done: true }], {
        type: "todo",
      });
      const tasks: TaskInstance[] = await paul.tasks.get(
        {
          title: { $sw: "B" },
          done: true,
        },
        { type: "todo" }
      );

      expect(
        tasks.map(t => {
          delete t.props.guid;
          return t;
        })
      ).toMatchSnapshot();
    });

    it("should find all tasks assigned 10 seconds ago", async () => {
      await paul.tasks.create(tasksProps);
      await paul.tasks.create([{ title: "Buy milk" }], {
        assigned: 1502022002035,
      });

      const tasks: TaskInstance[] = await paul.tasks.get(
        {},
        { assigned: { $gt: 1502022002035 - 10000 } }
      );

      expect(
        tasks.map(t => {
          delete t.props.guid;
          return t;
        })
      ).toMatchSnapshot();
    });

    it("should find all tasks assigned 10 seconds ago, type of todo and starts with 'B'", async () => {
      await paul.tasks.create(tasksProps);
      await paul.tasks.create(
        [{ title: "Buy milk" }, { title: "Read a book" }],
        {
          assigned: 1502022002035,
          type: "todo",
        }
      );

      const tasks: TaskInstance[] = await paul.tasks.get(
        { title: { $sw: "B" } },
        { assigned: { $gt: 1502022002035 - 10000 }, type: { $eq: "todo" } }
      );

      expect(
        tasks.map(t => {
          delete t.props.guid;
          return t;
        })
      ).toMatchSnapshot();
    });
  });

  describe("count", () => {
    let paul: PersonInstance;

    const tasksProps: TaskProps[] = [
      {
        title: "Buy milk",
      },
      {
        title: "Buy beer",
        done: true,
      },
    ];

    beforeEach(async () => {
      paul = await Person.create({ name: "Paul" });
      const hubert: PersonInstance = await Person.create({ name: "Hubert" });
      hubert.tasks.create([{ title: "Learn to drive a car" }]);
    });

    it("should count all tasks to instance", async () => {
      await paul.tasks.create(tasksProps);
      const tasks: number = await paul.tasks.count();
      expect(tasks).toMatchSnapshot();
    });

    it("should count all tasks with property done equals true to instance", async () => {
      await paul.tasks.create(tasksProps);
      const tasks: number = await paul.tasks.count({ done: true });
      expect(tasks).toMatchSnapshot();
    });

    it("should count all tasks which are assigned", async () => {
      await paul.tasks.create(tasksProps);
      await paul.tasks.create([{ title: "Buy milk" }], {
        assigned: 1502022002035,
      });
      const tasks: number = await paul.tasks.count(
        {},
        { assigned: { $gt: 0 } }
      );
      expect(tasks).toMatchSnapshot();
    });
  });

  describe("add", () => {
    const tasksProps: TaskProps[] = [
      {
        title: "Buy milk",
      },
      {
        title: "Buy beer",
        done: true,
      },
    ];

    it("should add only a relation of instances to instance", async () => {
      const paul: PersonInstance = await Person.create({ name: "Paul" });
      const tasks: TaskInstance[] = [];
      for (const props of tasksProps) {
        tasks.push(await Task.create(props));
      }
      const t: number = await paul.assignedTasks.add(tasks);
      expect(t).toMatchSnapshot();

      const relatedTasks = await paul.assignedTasks.get();

      const mapSort = t =>
        // $FlowFixMe
        t.map(a => a.props).sort((a, b) => a.title.localeCompare(b.title));
      expect(mapSort(relatedTasks)).toEqual(mapSort(tasks));
    });
  });

  describe("update", () => {
    const tasksProps: TaskProps[] = [
      {
        title: "Buy milk",
        done: false,
      },
      {
        title: "Buy beer",
        done: true,
      },
    ];

    it("should set tasks to done", async () => {
      const paul: PersonInstance = await Person.create({ name: "Paul" });
      const tasks: TaskInstance[] = await paul.tasks.create(tasksProps);
      const result: TaskInstance[] = await paul.tasks.update(
        { done: true },
        { done: false }
      );

      expect(
        result.map(r => {
          delete r.props.guid;
          return r;
        })
      ).toMatchSnapshot();
    });
  });
});
