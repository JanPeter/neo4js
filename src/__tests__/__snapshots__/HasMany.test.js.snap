// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HasMany add should add only a relation of instances to instance 1`] = `2`;

exports[`HasMany count should count all tasks to instance 1`] = `2`;

exports[`HasMany count should count all tasks with property done equals true to instance 1`] = `1`;

exports[`HasMany create should create a instances of tasks and relate them to person instance 1`] = `
Array [
  ModelInstance {
    "props": Object {
      "title": "Buy milk",
    },
  },
  ModelInstance {
    "props": Object {
      "done": false,
      "title": "Buy beer",
    },
  },
]
`;

exports[`HasMany get should find all tasks starting with character 'B' and are done 1`] = `
Array [
  ModelInstance {
    "props": Object {
      "done": true,
      "guid": "9f6504b5-dabd-4f11-bf97-f237be32b9e6",
      "title": "Buy beer",
    },
  },
]
`;

exports[`HasMany get should find all tasks to instance 1`] = `
Array [
  ModelInstance {
    "props": Object {
      "done": true,
      "title": "Buy beer",
    },
  },
  ModelInstance {
    "props": Object {
      "title": "Buy milk",
    },
  },
]
`;

exports[`HasMany get should find all tasks to instance 2`] = `
Array [
  ModelInstance {
    "props": Object {
      "done": true,
      "title": "Buy beer",
    },
  },
  ModelInstance {
    "props": Object {
      "title": "Buy milk",
    },
  },
]
`;

exports[`HasMany get should find all tasks to instance where task property done equals true 1`] = `
Array [
  ModelInstance {
    "props": Object {
      "done": true,
      "title": "Buy beer",
    },
  },
]
`;

exports[`HasMany get should find nothing 1`] = `Array []`;

exports[`HasMany update should set tasks to done 1`] = `
Array [
  ModelInstance {
    "props": Object {
      "done": true,
      "title": "Buy milk",
    },
  },
]
`;
